leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -137,34 +139,41 @@ public T buildResponse(final Response coapResponse) {
         endpoint.sendRequest(coapRequest);
     }
 
-    public synchronized void cancelPendingRequests(String registrationId) {
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
-        if (requests != null) {
-            for (Request coapRequest : requests) {
-                coapRequest.cancel();
-                removePendingRequest(registrationId, coapRequest);
-            }
-        }
+    private String getFloorKey(String registrationId) {
+        if (registrationId == null)
+            return null;
+        return registrationId + '#';
     }
 
-    private synchronized void addPendingRequest(String registrationId, Request coapRequest) {
-        // Get or create all pending requests for this registrationId
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
-        if (requests == null) {
-            requests = new ArrayList<>();
-            pendingRequests.put(registrationId, requests);
+    private String getCeilingKey(String registrationId) {
+        if (registrationId == null)
+            return null;
+        return registrationId + "#A";
+    }
+
+    private String getKey(String registrationId, Request coapRequest) {
+        if (registrationId == null || coapRequest == null)
+            return null;
+        return registrationId + '#' + coapRequest.hashCode();
```

Comment:
The java doc say :
```java
    /**
     * ....
     * As much as is reasonably practical, the hashCode method defined by
     * class {@code Object} does return distinct integers for distinct
     * objects. (This is typically implemented by converting the internal
     * address of the object into an integer, but this implementation
     * technique is not required by the
     * Java<font size="-2"><sup>TM</sup></font> programming language.)
```
But maybe this is not the best approach... I thought about the CoAP messageID  but not sure this is better?
The CoAP spec say that:
"The same Message ID MUST NOT be reused (in communicating with the same endpoint) within the EXCHANGE_LIFETIME"_